File: 40-combination-sum-ii/combination-sum-ii.cpp
class Solution {
public:
    void combinationSum(int idx , vector<int>&candidates , int tar, vector<vector<int>>&res, vector<int>&ans){
        if(tar == 0){
            res.push_back(ans);
            return;
        }
        for(int i = idx ; i < candidates.size() ; i++){
            if(i > idx && candidates[i] == candidates[i-1]) continue;
            if(candidates[i] > tar) break;
            ans.push_back(candidates[i]);
            combinationSum(i+1,candidates,tar - candidates[i],res,ans);
            ans.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        vector<vector<int>>res;
        vector<int>ans;
        combinationSum(0,candidates,target,res,ans);
        return res;
    }
};

File: 40-combination-sum-ii/combination-sum-ii.java
class Solution {
    public void creating(int[] candidates, int target, int sum, int ind, List<List<Integer>>res, List<Integer>lst){
        if(sum == target){
            res.add(new ArrayList<>(lst));
            return;
        }
        for(int i = ind ; i < candidates.length ; i++){
            if(i > ind && candidates[i] == candidates[i - 1]) continue;
            if(sum + candidates[i] > target) break;
            lst.add(candidates[i]);
            creating(candidates, target, sum + candidates[i], i + 1, res, lst);
            lst.remove(lst.size() - 1);
        }
    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>>res = new ArrayList<>();
        List<Integer>lst = new ArrayList<>();
        creating(candidates, target, 0 , 0 , res, lst);
        return res;
    }
}